### PR TITLE
Openssl 1.1.1 updates

### PR DESCRIPTION
**Pull request by MSP-Greg**
1. Updates tests and code for OpenSSL 1.1.1
2. Adds EM::OPENSSL_LIBRARY_VERSION and EM::OPENSSL_VERSION constants
3. Various updates to Travis for Ruby 2.6, change to xenial, adds MacOS, etc

### CODE REVIEW

**Comment by MSP-Greg** [comment]
Apologies, I should have added more comments to the commits...

**Comment by MSP-Greg** [comment]
I think I should split the above three commits into a few more (in particular the middle one).  Long story, the first two were done while OpenSSL 1.1.1 was still in beta.

Please give me a day or two...

File: .travis.yml [hunk 1]
```diff
@@ -1,25 +1,37 @@
...
+  - ruby-head
   - 2.6
-  - 2.5
-  - 2.4
-  - 2.3
-  - 2.2
+  - 2.5.3
+  - 2.4.5
+  - 2.3.8
+  - 2.2.10
```
Comment (sodabrew) [inline .travis.yml]: I prefer to let these versions float to latest - did you find unexpected versions/behavior without pinning these specifically?

File: .travis.yml [hunk 2]
```diff
@@ -1,25 +1,37 @@
...
+script:
+  - bundle exec rake compile test
+
 env:
   global:
-    - TESTOPTS=-v
+    - TESTOPTS="-v --no-show-detail-immediately --order=random"
+
 rvm:
+  - ruby-head
```
Comment (sodabrew) [inline .travis.yml]: ruby-head should be an allowed failure, though.

File: .travis.yml [hunk 3]
```diff
@@ -32,10 +44,20 @@ matrix:
...
 #    - rvm: jruby-9
   include:
-    - rvm: 2.3
+    - rvm: 2.6.0
+      os: osx
+    - rvm: 2.5.3
+      os: osx
+    - rvm: 2.4.5
+      os: osx
+    - rvm: 2.3.8
```
Comment (sodabrew) [inline .travis.yml]: I think it's too many versions on OS X - they're super slow to run anyways. Ideally the only version would be the system-Ruby on OS X, but I wasn't ever able to get that working in the Travis CI images.

File: Gemfile [hunk 4]
```diff
@@ -2,12 +2,7 @@ source 'https://rubygems.org'
 
 gemspec
 
-# Rake 11.0 no longer supports Ruby 1.8.7 and Ruby 1.9.2
-if RUBY_VERSION < '1.9.3'
-  gem 'rake', '< 11'
-else
-  gem 'rake'
-end
+gem 'rake'
```
Comment (sodabrew) [inline Gemfile]: 👍

File: ext/rubymain.cpp [hunk 5]
```diff
@@ -1569,5 +1569,33 @@ extern "C" void Init_rubyeventmachine()
...
+	rb_define_const (EmModule, "OPENSSL_NO_SSL2", Qfalse);
+#endif
+#endif
+
+  // OpenSSL Build / Runtime/Load versions
+
+	/* Version of OpenSSL that EventMachine was compiled with */
+	rb_define_const(EmModule, "OPENSSL_VERSION", rb_str_new2(OPENSSL_VERSION_TEXT));
+
+#if !defined(LIBRESSL_VERSION_NUMBER) && OPENSSL_VERSION_NUMBER >= 0x10100000
```
Comment (sodabrew) [inline ext/rubymain.cpp]: Thanks for including libressl!

File: tests/em_test_helper.rb [hunk 6]
```diff
@@ -4,11 +4,70 @@
...
+if EM.ssl?
+  require 'openssl'
+  ssl_lib_vers = OpenSSL.const_defined?(:OPENSSL_LIBRARY_VERSION) ?
+    OpenSSL::OPENSSL_LIBRARY_VERSION : 'na'
+  puts "OpenSSL OPENSSL_LIBRARY_VERSION: #{ssl_lib_vers}\n" \
+       "                OPENSSL_VERSION: #{OpenSSL::OPENSSL_VERSION}\n" \
+       "     EM OPENSSL_LIBRARY_VERSION: #{EM::OPENSSL_LIBRARY_VERSION}\n" \
+       "                OPENSSL_VERSION: #{EM::OPENSSL_VERSION}"
+
+  if RUBY_VERSION >= '2.0'
```
Comment (sodabrew) [inline tests/em_test_helper.rb]: Can remove this condition now.

File: tests/em_test_helper.rb [hunk 7]
```diff
@@ -4,11 +4,70 @@
...
+    SSL_AVAIL = temp.freeze
+  else
+    # Equal to base METHODS
+    SSL_AVAIL = ::OpenSSL::SSL::SSLContext::METHODS.select { |i| i =~ /[^\d]\d\z/ }.map { |i| i.to_s }
+  end
+
+  puts "                      SSL_AVAIL: #{SSL_AVAIL.sort.join(' ')}", ""
+end
+
+if RUBY_VERSION >= '2.0'
```
Comment (sodabrew) [inline tests/em_test_helper.rb]: Can remove this condition now.

File: tests/em_test_helper.rb [hunk 8]
```diff
@@ -4,11 +4,70 @@
...
+    
+    def teardown
+      EM.cleanup_machine
+    end
+  end
+end
 
 class Test::Unit::TestCase
-  class EMTestTimeout < StandardError ; end
+  if RUBY_VERSION >= '2.0'
```
Comment (sodabrew) [inline tests/em_test_helper.rb]: Can remove this condition now.

File: tests/test_iterator.rb [hunk 9]
```diff
@@ -52,8 +52,9 @@ def test_concurrency_bigger_than_list_size
         EM::Timer.new(1) {iter.next}
       }, proc {EM.stop})
     }
-    assert_equal(1, items.keys.size)
-    assert_equal(list.to_a.sort, items.values.flatten.sort)
+    assert(items.keys.size >= 1)
```
Comment (sodabrew) [inline tests/test_iterator.rb]: Why is this test now 1-or-more instead of exactly-1?

File: tests/test_iterator.rb [hunk 10]
```diff
@@ -74,7 +75,7 @@ def test_changing_concurrency_affects_active_iteration
         EM::Timer.new(0.2) {iter.next}
       }, proc {EM.stop})
     }
-    assert_in_delta(7, items.keys.size, 1)
+    assert_in_delta(7, items.keys.size, 2) # intermittent issues, 1 => 2
```
Comment (sodabrew) [inline tests/test_iterator.rb]: Oh, if this is why?

File: tests/test_ssl_ecdh_curve.rb [hunk 11]
```diff
@@ -27,18 +24,23 @@ def unbind
...
+        start_tls(:cipher_list => "ECDH", :ssl_version => %w(TLSv1_2))
+      else
+        start_tls(:ecdh_curve => "prime256v1", :cipher_list => "ECDH", :ssl_version => %w(TLSv1_2))
+      end
     end
 
     def ssl_handshake_completed
       $server_handshake_completed = true
       $server_cipher_name = get_cipher_name
+      close_connection if /TLSv1\.3/ =~ get_cipher_protocol
```
Comment (sodabrew) [inline tests/test_ssl_ecdh_curve.rb]: Isn't this match backwards?

File: tests/test_ssl_ecdh_curve.rb [hunk 12]
```diff
@@ -50,7 +52,7 @@ def ssl_handshake_completed
   def test_no_ecdh_curve
     omit("No SSL") unless EM.ssl?
     omit_if(rbx?)
-    omit("OpenSSL 1.1.x (and later) auto selects curve") if OPENSSL_GT_1_0
+    omit("OpenSSL 1.1.x (and later) auto selects curve") if SSL_LIB_VERS >= '1.1'
```
Comment (sodabrew) [inline tests/test_ssl_ecdh_curve.rb]: FYI - Nit - This is just a text comparison, not a formal version comparison. It's not a problem, just wanted to point it out.

File: tests/test_ssl_verify.rb [hunk 13]
```diff
@@ -1,74 +1,100 @@
...
 
     def ssl_handshake_completed
-      $client_handshake_completed = true
-      close_connection
+      @@handshake_completed = true
     end
 
     def unbind
-      EM.stop_event_loop
+      @@handshake_completed = false # if @is_tls1_3
```
Comment (sodabrew) [inline tests/test_ssl_verify.rb]: Is this `if` informational or accidentally commented-out?

File: tests/test_ssl_verify.rb [hunk 14]
```diff
@@ -1,74 +1,100 @@
...
     end
   end
 
   module Client
-    def connection_completed
-      start_tls(:private_key_file => $dir+'client.key', :cert_chain_file => $dir+'client.crt')
+    @@handshake_completed = nil
+    def self.handshake_completed? ; @@handshake_completed end
+    def self.timer=(val)          ; @@timer = true        end
```
Comment (sodabrew) [inline tests/test_ssl_verify.rb]: Does this intentionally ignore the `val` rvalue?

File: tests/test_ssl_verify.rb [hunk 15]
```diff
@@ -1,74 +1,100 @@
...
 
     def unbind
-      EM.stop_event_loop
+      @@handshake_completed = false unless @@timer
     end
   end
 
   module AcceptServer
+    @@handshake_completed = nil
+    def self.handshake_completed? ; @@handshake_completed end
```
Comment (sodabrew) [inline tests/test_ssl_verify.rb]: To make this a boolean, use `!!@@handshake_completed`

File: tests/test_ssl_verify.rb [hunk 16]
```diff
@@ -1,74 +1,100 @@
...
 
     def ssl_handshake_completed
-      $server_handshake_completed = true
+      @@handshake_completed = true
     end
   end
 
   module DenyServer
+    @@handshake_completed = nil
+    def self.handshake_completed? ; @@handshake_completed end
```
Comment (sodabrew) [inline tests/test_ssl_verify.rb]: As above.

File: tests/test_ssl_verify.rb [hunk 17]
```diff
@@ -1,74 +1,100 @@
...
 
     def ssl_handshake_completed
-      $server_handshake_completed = true
+      @@handshake_completed = true
     end
   end
 
   module FailServerNoPeerCert
+    @@handshake_completed = nil
+    def self.handshake_completed? ; @@handshake_completed end
```
Comment (sodabrew) [inline tests/test_ssl_verify.rb]: As above.

File: .travis.yml (hunk 1 above)
Comment (MSP-Greg) [reply .travis.yml]: Makes sense.  Will update.

**Comment by MSP-Greg** [comment]
@sodabrew

Thanks for looking at the code, as there's quite a bit.  Yesterday, I also reviewed.  I believe all of the items above are accounted for.

> I think it's too many versions on OS X - they're super slow to run

Re MacOS Ruby versions, 2.3.8 is using OpenSSL 1.0.2, 2.45 thru 2.6.0 are using 1.1.0.  Since xenial is using 1.0.2, I thought using MacOS to pick up 1.1.0 testing would be good (reminder, Appveyor Ruby 2.5 thru trunk are using OpenSSL 1.1.1).

Re 'time to run', I removed from `.travis.yaml`:
```
'if [[ "$TRAVIS_OS_NAME" = "osx" ]]; then brew install openssl; fi'
```

There have been issues with bundler, rubygems, what's Travis got loaded, etc.  Hence, rather than using 'bundle install & bundle exec', I tried the following:

```
before_install:
  #- 'if [[ "$TRAVIS_OS_NAME" = "osx" ]]; then brew install openssl; fi'
  # rubygems 2.7.8 and greater include bundler
  # remove 2.7.0 code when Travis removes rubygems 2.7.8 from ruby-head build
  - |
    rv="$(ruby -e 'STDOUT.write RUBY_VERSION')";
    if   [ "$rv" \< "2.3" ]; then gem update --system 2.7.8 --no-document
    elif [ "$rv" \< "2.7" ]; then gem update --system --no-document --conservative
    fi
  - gem install "rake:>=10.4.2" "rake-compiler:>=1.0.7" "minitest:>=5.4.3" "test-unit:>=3.2.9" --conservative --no-document

install: rake compile
script: rake test
```

Now, I'm not a MacOS type, so I don't know.  But after the change(s), the MacOS tests times are now about the same as the Ubuntu times.  See https://travis-ci.org/MSP-Greg/eventmachine/builds/478910088

I have a bias for always using the 'latest, greatest'.

**Comment by MSP-Greg** [comment]
@sodabrew

Decided to separate a commit for the console 'OpenSSL info' code, and I was wondering if the code to generate the following would be helpful as a diagnostic class method in EM?  It's currently in `tests/test_ssl_protocols.rb`, but it's moving to `tests/em_test_helper.rb`...

```
ruby 2.5.3p105 (2018-10-18 revision 65156) [x86_64-darwin17]

EM.library_type extension        EM.ssl? true
OpenSSL OPENSSL_LIBRARY_VERSION: OpenSSL 1.1.0h  27 Mar 2018
                OPENSSL_VERSION: OpenSSL 1.1.0h  27 Mar 2018
     EM OPENSSL_LIBRARY_VERSION: OpenSSL 1.1.0h  27 Mar 2018
                OPENSSL_VERSION: OpenSSL 1.1.0h  27 Mar 2018
                      SSL_AVAIL: TLSv1 TLSv1_1 TLSv1_2
```

**Comment by MSP-Greg** [comment]
@sodabrew

I think I've finished; sorry for taking so long.  I've got 6 commits in another branch.  Top six shown at https://github.com/MSP-Greg/eventmachine/commits/openssl-1.1.1-update.  Would you like the commits moved here, or just do another PR?  I believe I accounted for all your comments, along with some things I cleaned up...

I tried to get the commits to all pass sequentially for future bisect.  I believe Travis will work now, but Appveyor may fail on one or two due to OpenSSL 1.1.1.

Cut back on OSX, just running 2.3 & 2.6 (2.6 tests OpenSSL 1.1.0).  Travis test took under five minutes.

Stayed with bundler, but split it in two, for better 'folding' in the Travis log.  Also, left the 'OpenSSL' info in the test suite...

File: tests/test_ssl_ecdh_curve.rb (hunk 12 above)
Comment (MSP-Greg) [reply tests/test_ssl_ecdh_curve.rb]: Using an array now

**Comment by sodabrew** [comment]
Please do push the commits to this branch, so they'll appear in this same PR with the comments and discussion. Looking at the separate branch link, this is ready to land!

**Comment by MSP-Greg** [comment]
> Please do push the commits to this branch

Done.  Re Appveyor, my fork had one intermittent failure in Ruby 2.6 64, everything else passed.  Thanks.

**Pull request merged**